allennlp/layers/embeddings.py

```diff
@@ -0,0 +1,173 @@
+import codecs
+import gzip
+import logging
+
+import numpy
+import torch
+
+from ..data.vocabulary import Vocabulary
+from ..common.checks import ConfigurationError
+logger = logging.getLogger(__name__)  # pylint: disable=invalid-name
+
+
+class Embedding(torch.nn.Module):
+    """
+    A more featureful embedding module than the default in Pytorch.
+    Adds the ability to pre-specify the weight matrix, use a non-trainable embedding and
+    catches some ill-advised use cases, such as 1 dimensional embeddings.
+
+    Parameters
+    ----------
+    num_embeddings:, int:
+        Size of the dictionary of embeddings (vocabulary size).
+    embedding_dim: int
+        The size of each embedding vector.
+    weight: torch.FloatTensor, (optional, default=None)
+        A pre-initialised weight matrix for the embedding lookup, allowing the use of
+        pretrained vectors.
+    padding_index: int, (optional, default=None)
+        If given, pads the output with zeros whenever it encounters the index.
+    max_norm: float, (optional, default=None)
+        If given, will renormalize the embeddings to always have a norm lesser than this
+    norm_type: float, (optional, default=2):
+        The p of the p-norm to compute for the max_norm option
+    scale_grad_by_freq: boolean, (optional, default=False):
+        If given, this will scale gradients by the frequency of the words in the mini-batch.
+    sparse: bool, (optional, default=False):
+        Whether or not the Pytorch backend should use a sparse representation of the embedding weight.
+
+    Returns
+    -------
+    An Embedding module.
+
+    """
+
+    def __init__(self,
+                 num_embeddings: int,
+                 embedding_dim: int,
+                 weight: torch.FloatTensor=None,
+                 padding_index: int=None,
+                 trainable: bool=True,
+                 max_norm: float=None,
+                 norm_type: float=2.,
+                 scale_grad_by_freq: bool=False,
+                 sparse: bool=False):
+        super(Embedding, self).__init__()
+        self.num_embeddings = num_embeddings
+        self.embedding_dim = embedding_dim
+        self.padding_index = padding_index
+        self.max_norm = max_norm
+        self.norm_type = norm_type
+        self.scale_grad_by_freq = scale_grad_by_freq
+        self.sparse = sparse
+
+        if embedding_dim == 1:
+            raise ConfigurationError("There is no need to embed tokens if you "
+                                     "are using a 1 dimensional embedding.")
+
+        if weight is None:
+            weight = torch.FloatTensor(num_embeddings, embedding_dim)
+            self.weight = torch.nn.Parameter(weight, requires_grad=trainable)
+            self.reset_parameters()
```

Comment:
It looks like there is a technical reason to have this function if we ever need to call it to create parameters in the call of `forward`: https://discuss.pytorch.org/t/dynamic-parameter-declaration-in-forward-function/427/7

This isn't the reason it's used here though, but I was wondering why other pytorch layers use it and it looks like this might be why.